Implement critical section and critical region kernel functions

Fisherman166 opened the pull request:
Reverse engineered from the disassembled kernel functions dumped from my Xbox, these kernel functions should now match the original functionality found on the Xbox:

KeEnterCriticalRegion
KeLeaveCriticalRegion
RtlEnterCriticalSectionAndRegion
RtlEnterCriticalSection
RtlInitializeCriticalSection
RtlLeaveCriticalSectionAndRegion
RtlLeaveCriticalSection
RtlTryEnterCriticalSection

LukeUsher commented on src/CxbxKrnl/EmuKrnlKe.cpp:
```diff
@@ -563,19 +563,29 @@ XBSYSAPI EXPORTNUM(100) xboxkrnl::VOID NTAPI xboxkrnl::KeDisconnectInterrupt
+// Many Xbox Kernel functions load this pointer and make changes to fields
+// in it. More research will be needed to define all fields
+uint8_t* get_thread_Teb() {
+    return (uint8_t*)0x80047BF0;
```
The Teb is part of the Thread information data structure contained within the PCR. You can call KeGetPcr() from any thread to get a pointer to the emulated KPCR, then you can access it as a struct and modify/update the necessary fields.

LukeUsher replied:
See https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/blob/c8c0353374e16f5482b4188ea00678a2b0a9876b/import/OpenXDK/include/xboxkrnl/xboxkrnl.h#L2026 for the layout of the _KPRCB structure, which is stored within the _KPCR structure along with the Tib

https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/blob/c8c0353374e16f5482b4188ea00678a2b0a9876b/import/OpenXDK/include/xboxkrnl/xboxkrnl.h#L2048

LukeUsher commented on src/CxbxKrnl/EmuKrnlKe.cpp:
```diff
@@ -563,19 +563,29 @@ XBSYSAPI EXPORTNUM(100) xboxkrnl::VOID NTAPI xboxkrnl::KeDisconnectInterrupt
+uint8_t* get_thread_Teb() {
+    return (uint8_t*)0x80047BF0;
+}
+#define CRITICAL_REGION_UNKNOWN1 (get_thread_Teb() + 0x34)
```
Please use KeGetPcr and the KPCR struct to access these values

LukeUsher commented on src/CxbxKrnl/EmuKrnlKe.cpp:
```diff
@@ -1037,14 +1047,22 @@ void ConnectKeInterruptTimeToThunkTable()
+    uint32_t* critical_region_unk3 = (uint32_t*)CRITICAL_REGION_UNKNOWN3;
 
-	LOG_UNIMPLEMENTED();
+    *critical_region_unk3++;
```
this is KThread->KernelApcDisable, You can get the KThread object by calling KeGetCurrentThread.

LukeUsher commented on src/CxbxKrnl/EmuKrnlKe.cpp:
```diff
@@ -1037,14 +1047,22 @@ void ConnectKeInterruptTimeToThunkTable()
+    *critical_region_unk3++;
+    if(*critical_region_unk3 == 0) {
+        if(*critcal_region_unk1 != (uint32_t)critcal_region_unk1) {
+            *critcal_region_unk2 = 1;
```
unk1 is KThread->ApcState->KernelApcPending

LukeUsher commented on src/CxbxKrnl/EmuKrnlKe.cpp:
```diff
@@ -1037,14 +1047,22 @@ void ConnectKeInterruptTimeToThunkTable()
-	LOG_UNIMPLEMENTED();
+    *critical_region_unk3++;
+    if(*critical_region_unk3 == 0) {
+        if(*critcal_region_unk1 != (uint32_t)critcal_region_unk1) {
```
Unk1 is KThread->ApcState->ApcListHead

LukeUsher commented on src/CxbxKrnl/EmuKrnlRtl.cpp:
```diff
@@ -971,33 +905,30 @@ XBSYSAPI EXPORTNUM(306) xboxkrnl::BOOLEAN NTAPI xboxkrnl::RtlTryEnterCriticalSec
+    #endif
+
+    if(exchange_result == -1) {
+        CriticalSection->OwningThread = (HANDLE)get_thread_Teb();
```
`CriticalSection->OwningThread = (HANDLE)KeGetCurrentThread();`

LukeUsher commented on src/CxbxKrnl/EmuKrnlRtl.cpp:
```diff
@@ -971,33 +905,30 @@ XBSYSAPI EXPORTNUM(306) xboxkrnl::BOOLEAN NTAPI xboxkrnl::RtlTryEnterCriticalSec
+        ret = true;
+    }
+    else {
+        if(CriticalSection->OwningThread == (HANDLE)get_thread_Teb()) {
```
Again, this should use KeGetCurrentThread()

PatrickvL commented:
Judging from all comments it's clear we're all very interested in getting this right. So, thanks for this, I think it'll be stirring up a discussion.

haxar commented:
There is no need to reverse engineer these calls. They're pretty well documented.

PatrickvL commented:
@haxar Actually, I very much appreciate @Fisherman166 took the trouble to reverse-engineer this. Having & improving these skills is critically important in implementing other parts of our kernel, so it's beneficial for us to keep receiving contributions like this (regardless the current amount of review-remarks - those will surely reduce in time).

Fisherman166 commented:
Sorry I dropped the ball guys. My Vim on windows was setup case insensitive so when I searched for 'currentthread' nothing came up. A grep would have solved that issue of course....And then I also didn't have my ctags setup to tag the import/ folder so I missed all of those data structures. Finding this stuff 2 days ago would have saved me a lot of headache :)

I added all of the fixes requested in this thread and pushed them. Patrick did bring up a good point though that the host OS no longer has any idea that this is occurring since I removed the Ntdll critical section code. My mental model of the emulator was lacking in that regard which may not make this update as great as I had hoped.

PatrickvL commented:
Closed without merging, in agreement after discussing the problem this change brings : Host thread scheduling won't wait on these Xbox critical sections.
